dashboard/app/controllers/activities_controller.rb

```diff
@@ -194,6 +194,13 @@ def track_progress_for_user
       end
     end
 
+    is_csf_course = @script_level.try(:script).try(:curriculum_umbrella) == 'CSF'
+    is_free_play = @level.try(:free_play) ? @level.try(:free_play).to_bool : false
+    is_validated_level = Level::TYPES_WITH_IDEAL_LEVEL_SOURCE.include?(@level.type) && @level.try(:ideal_level_source)
+    if is_csf_course && is_validated_level && !is_free_play
+      record_time_spent(@user_level)
+    end
```

Comment:
All the `.try` clauses make me thing we haven't put this code in the right spot.  I wonder if it belongs somewhere down in `User.track_level_progress` since we know that will always happen if this does.

Reply:
This is indeed some complex application logic to add on top of an already-complex controller-action. Here are some refactoring ideas (just suggestions, nothing necessary) to [prefer skinny controllers](https://github.com/code-dot-org/code-dot-org/blob/staging/STYLEGUIDE.md#rails) and move/simplify some of this new logic to more modular methods on the corresponding models:

- Since `record_time_spent` encapsulates a data operation performed on UserLevel (and associated ValidatedUserLevel), it could be moved from the controller to a method on the UserLevel model to simplify some of the `try` chains.
- `is_csf_course` can reuse the existing method on the `Script` model (`script.csf?`)
- `is_validated_level` can be extracted to a method on the `Level` model (`level.validated?`)
- The combined csf-validated logic can be extracted to a method on the `user_level` model (`user_level.csf_validated?`)

- Finally, you can use a [Rails Association](https://guides.rubyonrails.org/association_basics.html) [`has_one`](https://guides.rubyonrails.org/association_basics.html#has-one-association-reference), [`create_[association]`](https://guides.rubyonrails.org/association_basics.html#methods-added-by-has-one-create-association-attributes), and [`#increment!`](https://api.rubyonrails.org/classes/ActiveRecord/Persistence.html#method-i-increment-21) to simplify some of the association-creation and attribute-increment logic in `record_time_spent`.

Put it all together and it would look something like this (just a sketch, I haven't tested anything):

```ruby
# track_progress_for_user
time_spent = params[:timeSinceLastMilestone].to_i, MAX_INT_MILESTONE].min
@user_level&.record_time_spent(time_spent)

# Level
def validated?
  TYPES_WITH_IDEAL_LEVEL_SOURCE.include?(type) && ideal_level_source && try(:free_play?)
end

# UserLevel
has_one :validated_user_level

def csf_validated?
  Script.get_from_cache(script_id).csf? && Script.cache_find_level(level_id).validated?
end

def record_time_spent(time)
  return unless time > 0 && csf_validated?
  Retryable.retryable on: [Mysql2::Error, ActiveRecord::RecordNotUnique], matching: /Duplicate entry/ do
    (validated_user_level || create_validated_user_level).increment!(time)
  end
end
```

Reply:
Thanks for the suggestions Will! I learned something new which was great! I implemented most of your suggestions but did notice that `increment!` looks to be blacklisted by our robocop here (https://github.com/code-dot-org/code-dot-org/blob/staging/.rubocop_todo.yml#L199). I think its because when I looked it up it says it does not validate the change before saving.